copyDesktopItems: add new setup-hook

B4dM4n opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

A change made to the `makeDesktopItem` helper introduced by #105099 broke the build of some packages.

These packages used the `desktopItem.buildCommand` string to install their desktop files directly, instead of using the `desktopItem` output:

https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/applications/audio/goattracker/default.nix#L56
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/applications/networking/mailreaders/thunderbird/68.nix#L291
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/applications/networking/mailreaders/thunderbird/default.nix#L286
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/applications/office/softmaker/generic.nix#L114-L116
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/games/minecraft/default.nix#L114
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/games/xonotic/default.nix#L154
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/tools/misc/jdiskreport/default.nix#L41
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/tools/security/jd-gui/default.nix#L95

This PR adds the `copyDesktopItems` setup hook, which simplifies the process of adding desktop files to a package. It is invoked as a `postInstallHook` and will install the files given to the `desktopItems` attribute into the `out` output.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

CC: @flokli @samueldr

B4dM4n committed:
copyDesktopItems: add new setup-hook

B4dM4n committed:
goattracker: use copyDesktopItems hook

B4dM4n committed:
thunderbird: use copyDesktopItems hook

B4dM4n committed:
softmaker-office: use copyDesktopItems hook

B4dM4n committed:
minecraft: use copyDesktopItems hook

B4dM4n committed:
xonotic: use copyDesktopItems hook

B4dM4n committed:
jdiskreport: use copyDesktopItems hook

B4dM4n committed:
jd-gui: use copyDesktopItems hook

SuperSandro2000 commented on pkgs/build-support/setup-hooks/copy-desktop-items.sh:
```diff
@@ -0,0 +1,42 @@
+postInstallHooks+=(copyDesktopItems)
+
+copyDesktopItems() {
+    if [ "${dontCopyDesktopItems-}" = 1 ]; then return; fi
```
```suggestion
    [[ ${dontCopyDesktopItems-} == 1 ]] && return
```
Normally in bash I would write the following but this looks more like sh or dash but you told shellcheck it is bash. I am a bit confused. Would you mind clearing this up?

B4dM4n replied:
I took this syntax directly from another hook. There are many which use the same style, so I copied it:

https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/build-support/setup-hooks/move-lib64.sh#L11
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/build-support/setup-hooks/move-sbin.sh#L8
https://github.com/NixOS/nixpkgs/blob/05d408e639f03be0c5db6e5c0d6621081f7106f2/pkgs/build-support/setup-hooks/update-autotools-gnu-config-scripts.sh#L4

I'm currently unaware if there is any policy about the hook sh/bash syntax, so I tried to stick with the existing hook style.

petabyteboy replied:
I'm pretty sure both are valid bash. Isn't bash a superset of sh anyways?

Ericson2314 replied:
Use `[[` .. `]]` it's better behaved. And when in doubt see what shellcheck says hehe (though I think it will instruct on both).

petabyteboy commented:
I guess the alternative would be to change all derivations using the old pattern to just copy the files using `install`, `ln`, or `cp` as some other derivations already do. I do like this solution more.
Either way if there are no new concerns I would like to merge soon since many of the applications I use are broken on master and fixed by this change.

petabyteboy commented:
Built minecraft, xonotic-sdl, thunderbird, jd-gui, softmaker-office, goattracker on x86_64-linux and confirmed the .desktop files were included in the output.

I rebased with these three changes:
- fixed conflict with latest master (minecraft package was bumped in
- minecraft was

Ericson2314 commented:
Thanks for cleaning this up! The old way has bugged me too.

Besides the comment above, can you perhaps find a way to share some code with the old `makeDesktopItem`? (Or explain why that is no good?)

FRidh commented:
Because https://github.com/NixOS/nixpkgs/issues/103071 broke the build of thunderbird, and this resolves it, I am going to merge this, even do some improvements have been suggested. It would be nice to see those in a follow-up commit.

B4dM4n commented:
> I rebased with these three changes:
> 
> * fixed conflict with latest master (minecraft package was changed in master)
> * minecraft was missing a `desktopItems = [ desktopItem ];`
> * jd-gui was missing a `desktopItems = [ desktopItem ];`

@petabyteboy  Thanks for fixing this.

> Besides the comment above, can you perhaps find a way to share some code with the old `makeDesktopItem`? (Or explain why do that is no good?)

There is nothing wrong with `makeDesktopItem` and all packages, besides the ones mentioned above, use it correctly by copying or linking the produced output. It's the usage of the `.buildCommand` attribute, which is the input of the `makeDesktopItem` derivation, as a shell input of another derivation (which now lacks the required `nativeBuildInputs`).

This should have never been allowed in the first place. (Question to self: Why are the inputs to `derivation(...)` available on the output?)

As (hopefully) all the bad uses of `makeDesktopItem` are now removed from nixpkgs, I don't see a reason to document exactly this usage, since all other examples in nixpkgs that people can use yield a correct result.

Documenting the hook in general might be a good idea.

Ericson2314 commented:
> There is nothing wrong with makeDesktopItem and all packages, besides the ones mentioned above, use it correctly by copying or linking the produced output. It's the usage of the .buildCommand attribute, which is the input of the makeDesktopItem derivation, as a shell input of another derivation (which now lacks the required nativeBuildInputs).

I agree with that. But I don't think the fact that `makeDesktopItem` isn't bad is a reason not share code?

> Question to self: Why are the inputs to derivation(...) available on the output?

Nix choose to implement the `builtins.derivation` primop that way, but I agree it is dubious.

B4dM4n commented:
Did you envision something tike this faee4397a10cf53fbe5f5741c91724623a23b10c?

https://github.com/NixOS/nixpkgs/blob/faee4397a10cf53fbe5f5741c91724623a23b10c/pkgs%2Fbuild-support%2Fmake-desktopitem%2Fdefault.nix#L51-L58

This would allow  outputs of `makeDesktopItem` to be used as `buildInputs` in other derivations. (example 59076059a9c7de1176c83079237187cfc802c5e7)

Ericson2314 commented:
@B4dM4n OK I guess there are less duplicated than I thought.

OK first a question. When would one use `makeDesktopItem` without `copyDesktopItems`? I can't think of any real use-case.

I suppose `makeDesktopItem` could just put the result in `$out` not `$out/share/applications` to hint that it's up to `copyDesktopItems` to install them in the right right.

B4dM4n commented:
Nearly all of the uses of `makeDesktopItem` in nixpkgs copy or link the desktop files. The only other use is in documentation module:

https://github.com/NixOS/nixpkgs/blob/9a01e978244c2d7e20b97533309f02ba93933a8c/nixos%2Fmodules%2Fmisc%2Fdocumentation.nix#L62-L77

Using `makeDesktopItem` to create directly installable or symlinkable packages which doesn't use `copyDesktopItems` is a nice feature in that case.



> I suppose `makeDesktopItem` could just put the result in `$out` not `$out/share/applications` to hint that it's up to `copyDesktopItems` to install them in the right place.

This sounds feasable. `copyDesktopItems` can use `stripHash` to get the target filename directly from `$out`. For uses of `makeDesktopItem` without `copyDesktopItems`, a attrbiute could be added to use `$out/share/applications` instead of `$out`.

Changing the default to `$out` would sadly break all uses of `makeDesktopItem` outside of nixpkgs.

Ericson2314 commented:
Ah `symlinkJoin` is a good enough reason to keep with `$out/share/applications`. OK, no more issues from me!
